code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -319,19 +319,78 @@
 
 /obj/item/borg/upgrade/selfrepair
 	name = "self-repair module"
-	desc = "This module will repair the cyborg over time."
+	desc = "This module will provide rapid repairs, provided it has had time to charge since its last use"
 	icon_state = "cyborg_upgrade5"
 	require_module = 1
-	var/repair_amount = -1
-	/// world.time of next repair
-	var/next_repair = 0
-	/// Minimum time between repairs in seconds
-	var/repair_cooldown = 4
-	var/msg_cooldown = 0
-	var/on = FALSE
-	var/powercost = 10
+
+	var/mutable_appearance/timer_overlay  // This entire block is used for the cooldown overlay
+	var/mutable_appearance/text_overlay
+	var/timer_overlay_active = FALSE
+	var/timer_icon = 'icons/effects/cooldown.dmi'
+	var/timer_icon_state_active = "second"
+	COOLDOWN_DECLARE(recharging_repmod)
+
+	/// how many times the repair tries to tick.
+	var/repair_ticks = 10
+	/// amount repaired per tick
+	var/repair_amount = -5
+	/// Power cell cost per tick. Repair aborts early if there is not at least 10x this much remaining at the start of a tick in order to preserve some power.
+	var/powercost = 50		
 	var/mob/living/silicon/robot/cyborg
-	var/datum/action/toggle_action
+	var/datum/action/action
+	var/activation_sound = 'sound/machines/terminal_processing.ogg'
+	var/running_sounds = list('sound/machines/generator/generator_mid1.ogg', 'sound/machines/generator/generator_mid2.ogg', 'sound/machines/generator/generator_mid3.ogg')
+	var/deactivation_sound = 'sound/effects/turbolift/turbolift-close.ogg'
+
+/obj/item/borg/upgrade/selfrepair/proc/begin_timer_animation()
+	if(!(action?.button) || timer_overlay_active)
+		return
+
+	timer_overlay_active = TRUE
+	timer_overlay = mutable_appearance(timer_icon, timer_icon_state_active)
+	timer_overlay.alpha = 180
+
+	if(!text_overlay)
+		text_overlay = image(loc = action.button, layer=ABOVE_HUD_LAYER)
+		text_overlay.maptext_width = 64
+		text_overlay.maptext_height = 64
+		text_overlay.maptext_x = -8
+		text_overlay.maptext_y = -6
+		text_overlay.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
+
+	if(action.owner?.client)
+		action.owner.client.images += text_overlay
+
+	action.button.add_overlay(timer_overlay, TRUE)
+	action.has_cooldown_timer = TRUE
+	update_timer_animation()
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/item/borg/upgrade/selfrepair/proc/update_timer_animation()
+	if(!(action?.button))
+		return
+	text_overlay.maptext = "<center><span class='chatOverhead' style='font-weight: bold;color: #eeeeee;'>[FLOOR(COOLDOWN_TIMELEFT(src, recharging_repmod)/10, 1)]</span></center>"
+
+/obj/item/borg/upgrade/selfrepair/proc/end_timer_animation()
+	if(!(action?.button) || !timer_overlay_active)
+		return
+	timer_overlay_active = FALSE
+	if(action.owner?.client)
+		action.owner.client.images -= text_overlay
+	action.button.cut_overlay(timer_overlay, TRUE)
+	timer_overlay = null
+	qdel(text_overlay)
+	text_overlay = null
+	action.has_cooldown_timer = FALSE
+
+	STOP_PROCESSING(SSfastprocess, src)
```

Comment:
I hate that this is just dupilcated spell code. The cooldown needs to be generalised to be accessible on non-spell actions, or this needs to be a spell (which sounds a bit weird).